### ISSUE TITLE
[Tracking] E2E training alignment and convergence against Megatron/Megatron-Bridge

### ISSUE THREAD

**Issue by HuiyingLi** [NVIDIA-NeMo/Automodel#3034]
We should add an end-to-end training alignment test suite that compares AutoModel against Megatron/Megatron-Bridge under identical settings. The goal is to catch subtle correctness issues in the full training path, including tokenizer/chat rendering, sample packing, answer-only loss masks, reasoning-content masks, position ids, forward numerics, backward numerics, and optimizer-step behavior.

Recent answer-only loss-mask issues in multiturn chat rendering show that component-level tests are not enough. Some bugs may not reliably surface as obvious training failures, but can still change the effective objective or training dynamics. We need a reference-based alignment workflow for representative model families.

## Scope

Compare AutoModel training against Megatron/Megatron-Bridge for these model families:

- Qwen3
- Qwen3.5
- Gemma4
- DeepSeek v4
- GLM 5.2

The test should cover the full training process, not only chat template rendering.

**Comment by khazic** [comment]
good job

**Comment by HuiyingLi** [comment]
## Gemma 4 31B convergence update: AutoModel vs Megatron Bridge

I completed a 100-step Gemma 4 31B pretraining convergence comparison between AutoModel (AM) and Megatron Bridge (MB).

**Scope note:** this validates the random-init pretraining path through forward, backward, optimizer step, scheduler, and checkpoint/resume. It does not yet cover every dimension proposed in this issue, such as chat rendering, answer-only/reasoning masks, or sample packing, and it is not yet a reusable CI test.

### Frozen comparison contract

| Field | Value |
|---|---|
| Model | `google/gemma-4-31B-it`, architecture/config only |
| Initialization | Random initialization, seed 1234 |
| Dataset | Identical pretokenized RedPajama2 indexed token IDs |
| Sequence length | 4096 |
| Global / micro batch | 1024 / 1 |
| Optimizer steps | 100 |
| Token slots | 4,194,304 per step; 419,430,400 total |
| Precision | BF16 model/compute; FP32 master parameters and Adam moments |
| Parallelism | TP4, PP1, CP1, EP1 |
| LR schedule | 40-step warmup to `3e-4`, cosine decay to `3e-5` |
| Adam | β1=0.9, β2=0.95, ε=`1e-8` |
| Gradient clipping | 1.0 |
| Gemma semantics | vocab 262144; final-logit softcap 30.0 |
| Activation checkpointing | All 60 trainable text layers |

Runtime audits verified the model initialization and materialized FP32 master parameters, first moments, and second moments. Dataset and model-config hashes were also recorded in W&B.

### Functional architecture parity

AutoModel has 30,697,345,280 trainable text parameters. Megatron allocates 30,807,445,760 parameters because its fused QKV tensor retains an unused V slice in each of the ten global-attention layers:

```text
10 × 5376 × (4 × 512) = 110,100,480
```

Gemma 4 uses `attention_k_eq_v`, so these layers reuse K as V. The extra MB allocation is not part of the functional forward path; both sides have the same 30,697,345,280 functional text parameters.

### Convergence results

| Measurement | AutoModel | Megatron Bridge |
|---|---:|---:|
| Initial loss | 14.22843266 | 14.19900894 |
| First-10 mean loss | 14.20518532 | 14.19064932 |
| Last-10 mean loss | **6.50286016** | 6.66565032 |
| Final loss | **6.44232941** | 6.62480783 |
| Whole-run loss slope / step | -0.09272592 | -0.10287208 |
| Minimum observed loss | 6.44232941 | 6.62480783 |
| Maximum observed loss | 17.08747101 | 21.95310211 |

Both jobs completed exactly 100 finite optimizer steps with no NaN or skipped iterations. Both first-to-last windows decreased substantially, and both whole-run slopes are negative. The serialized LR histories match to within `1.44e-11`.

Paired loss metrics:

| Metric | Value |
|---|---:|
| Final absolute loss difference | 0.18247843 |
| Last-10 mean loss difference | 0.16279016 |
| 100-step loss MAE | 0.74655903 |
| 100-step loss RMSE | 1.45924348 |
| Maximum absolute stepwise difference | 8.94609451 |

The larger stepwise differences are concentrated in the early transient phase. The late-training losses are much closer.

### Interpretation

This supports **matched qualitative convergence** for the Gemma 4 31B pretraining path:

- both implementations learn under the same data, initialization, optimizer, schedule, and functional architecture;
- both produce stable, finite training trajectories;
- both reach a similar late-training loss range.

This does **not** establish stepwise numerical identity. Different kernels, reduction orders, tensor layouts, and distributed execution paths produce expected numerical divergence. This is also one run per implementation and therefore does not measure run-to-run variance.

The fail-closed W&B pair audit records:

```text
comparison/qualitative_convergence_valid = 1
comparison/schedule_valid = 1
comparison/valid = 1
```

A separate checkpoint/resume gate also verified continuity of model, optimizer, scheduler, RNG, and dataloader state.

### W&B evidence

- [Convergence overlay report](https://wandb.ai/Nemo-automodel/automodel_convergence_runs/reports/Gemma-4-31B-AutoModel-vs-Megatron-convergence-overlay--VmlldzoxNzY0MjgzMw==)
- [AutoModel native run `19mnq6cb`](https://wandb.ai/Nemo-automodel/automodel_convergence_runs/runs/19mnq6cb)
- [Megatron Bridge native run `ftsox54u`](https://wandb.ai/Nemo-automodel/automodel_convergence_runs/runs/ftsox54u)
- [Paired convergence audit `doimm0gj`](https://wandb.ai/Nemo-automodel/automodel_convergence_runs/runs/doimm0gj)
- [Normalized AM history `z9soawev`](https://wandb.ai/Nemo-automodel/automodel_convergence_runs/runs/z9soawev)
- [Normalized MB history `3oxny091`](https://wandb.ai/Nemo-automodel/automodel_convergence_runs/runs/3oxny091)
- Checkpoint/resume proof: [reference `g4am14953036r`](https://wandb.ai/Nemo-automodel/automodel_convergence_runs/runs/g4am14953036r), [resumed `g4am14953036s`](https://wandb.ai/Nemo-automodel/automodel_convergence_runs/runs/g4am14953036s), and [audit `g4am14953036a`](https://wandb.ai/Nemo-automodel/automodel_convergence_runs/runs/g4am14953036a)

This result can serve as the Gemma 4 convergence baseline when the workflow is converted into the reusable E2E alignment suite proposed here.

**Comment by HuiyingLi** [comment]
## Qwen3-30B-A3B AutoModel vs. Megatron Bridge convergence verification

### Result

**PASS after [PR #3359 — `fix: normalize MoE auxiliary loss during gradient accumulation`](https://github.com/NVIDIA-NeMo/Automodel/pull/3359).**

A fresh bounded 100-step AutoModel run with the fix closely follows the Megatron Bridge reference curve:

* Fixed AutoModel final loss: `6.0536108`
* Fresh Bridge final loss: `6.0938020`
* Absolute difference: `0.0401912`
* Relative difference versus Bridge: `0.66%`

This report covers convergence only. Checkpoint/resume and performance comparisons are outside its scope.

### Verification contract

* Model: `Qwen/Qwen3-30B-A3B`, random initialization and full-model training
* Data: bounded RedPajama2 `head_01`
* Sequence length: 4096
* Global batch size / microbatch size: 1024 / 1
* Execution: 16 H100 GPUs, DP16, EP16, TP1/PP1/CP1, 64 gradient-accumulation microbatches
* Tokens: 4,194,304 per optimizer update; 419,430,400 total
* Updates: 100
* LR: linear warmup from zero to `3e-4` over 40 updates, then cosine decay to `3e-5`
* Optimizer: Adam, β=(0.9, 0.95), ε=`1e-8`, effective constant weight decay `0.033`, gradient clipping `1.0`
* Precision: BF16 model/compute with FP32 master parameters and FP32 optimizer moments
* Seed: 1234
* Routing: natural Qwen3-MoE routing, without forced balancing or token dropping

### Original divergence

The original AutoModel run was numerically stable but converged to a materially higher endpoint:

| Run | Initial loss | Final loss | Final-10 mean loss |
|---|---:|---:|---:|
| Original AutoModel | 12.4135485 | 7.3648901 | 7.3690899 |
| Fresh Megatron Bridge | 12.4114500 | 6.0938020 | 6.1232745 |
| Published Bridge card | 12.4114500 | 6.1391160 | — |

The close initial losses showed that model/data initialization was aligned, while the growing training gap pointed to an optimizer-step or backward-scaling difference. Runtime audits confirmed that AutoModel materialized FP32 master parameters and FP32 optimizer moments, so missing FP32 optimizer weights were **not** the cause.

### Root cause

The main language-model loss was normalized over the effective global batch, but model-provided MoE auxiliary losses entered every microbatch backward pass without the matching gradient-accumulation normalization.

For this run, one optimizer update contains 64 microbatches. The auxiliary gradient was therefore accumulated with materially excessive weight relative to the normalized language-model objective, progressively changing the optimization trajectory even though the first-step losses matched.

### Fix in PR #3359

PR #3359 centralizes the runtime-aware auxiliary-loss scale in `BaseRecipe._set_moe_aux_loss_backward_scale`.

The shared scaling rule is:

```
model_microbatches = outer_batches × pipeline_schedule_microbatches
scale = context_parallel_size / model_microbatches
```

When pipeline parallelism performs post-backward token normalization, the helper also applies the corresponding label-token and DP×CP compensation.

For the verified topology (PP1, CP1, 64 accumulated microbatches), the relevant per-microbatch auxiliary scale is `1/64`.

The PR:

* applies one topology-aware scaling path across LLM training, VLM finetuning, LLM KD, and VLM KD;
* accounts for gradient accumulation, context parallelism, and pipeline-parallel token normalization;
* removes duplicated recipe-local scaling logic;
* adds a reference optimizer-step test comparing distributed DP/CP/PP auxiliary gradients with a single-process FP32 objective;
* extends recipe tests for non-PP accumulation, CP compensation, PP schedule microbatches, zero-label batches, VLM, and KD paths.

Changed implementation paths:

* `nemo_automodel/recipes/base_recipe.py`
* `nemo_automodel/recipes/llm/train_ft.py`
* `nemo_automodel/recipes/llm/kd.py`
* `nemo_automodel/recipes/vlm/finetune.py`
* `nemo_automodel/recipes/vlm/kd.py`

### Fixed convergence result

| Run | Initial loss | Final loss | Final-10 mean loss |
|---|---:|---:|---:|
| Fixed AutoModel | 12.4092541 | 6.0536108 | 6.0881136 |
| Fresh Megatron Bridge | 12.4114500 | 6.0938020 | 6.1232745 |
| Published Bridge card | 12.4114500 | 6.1391160 | — |

Fixed AutoModel versus the fresh Bridge reference:

* Final absolute loss difference: `0.0401912`
* Final relative difference: `0.66%`
* Final-10 mean absolute difference: `0.0351609`
* All 100 loss, LR, and grad-norm rows were finite
* Exactly 4,194,304 label tokens were consumed per update
* No skipped or NaN optimizer iterations were observed

The fixed AutoModel curve tracks the Bridge curve and no longer exhibits the original convergence divergence. The remaining endpoint difference is within the expected cross-framework execution fingerprint; this is a convergence-alignment result, not a bitwise-equivalence claim.

### Validation

* Focused auxiliary-loss tests: **23 passed**
* Relevant recipe/training regression suite: **226 passed, 12 skipped**
* Ruff: clean
* Import-linter: clean
* Fresh 16-H100, 100-step AutoModel run: completed successfully

### Evidence

* Fixed AutoModel Slurm job: `14882842`
* Fresh Bridge Slurm job: `14879085`
* Fixed-run AutoModel source commit: `67d4887a6e1b864fd7a371d6fedb8a1d7305db43`
* [W&B loss/LR overlay](https://wandb.ai/Nemo-automodel/automodel_convergence_runs/reports/AMINT-229-Qwen3-30B-A3B-convergence-overlay--VmlldzoxNzYzODA4Ng==)

<img width="646" height="230" alt="Image" src="https://github.com/user-attachments/assets/7ea6a6b9-ecee-4cec-bd9c-cf380a161c44" />

* [Fix PR #3359](https://github.com/NVIDIA-NeMo/Automodel/pull/3359)

### Conclusion

The Qwen3-30B-A3B bounded convergence gap was caused by missing MoE auxiliary-loss normalization across gradient-accumulation microbatches. PR #3359 corrects the scaling for the supported distributed topologies and restores convergence alignment with Megatron Bridge.
